zenoh-ts/src/query.ts

```diff
@@ -0,0 +1,574 @@
+import { ReplyErrorWS } from "./remote_api/interface/ReplyErrorWS";
+import { UUIDv4 } from "./remote_api/session";
+import { QueryWS } from "./remote_api/interface/QueryWS";
+import { QueryReplyWS } from "./remote_api/interface/QueryReplyWS";
+// API
+import { IntoKeyExpr, KeyExpr } from "./key_expr";
+import { IntoZBytes, ZBytes } from "./z_bytes";
+import { Sample, Sample_from_SampleWS } from "./sample";
+import { Encoding } from "./encoding";
+
+//  ██████  ██    ██ ███████ ██████  ██    ██  █████  ██████  ██      ███████
+// ██    ██ ██    ██ ██      ██   ██  ██  ██  ██   ██ ██   ██ ██      ██
+// ██    ██ ██    ██ █████   ██████    ████   ███████ ██████  ██      █████
+// ██ ▄▄ ██ ██    ██ ██      ██   ██    ██    ██   ██ ██   ██ ██      ██
+//  ██████   ██████  ███████ ██   ██    ██    ██   ██ ██████  ███████ ███████
+//     ▀▀
+
+/**
+ * Queryable class used to receive Query's from the network and handle Reply's
+ * created by Session.declare_queryable
+ */
+export class Queryable {
+  private _remote_queryable: RemoteQueryable;
+  private _callback_queryable;
+  /** Finalization registry used for cleanup on drop
+   * @hidden 
+   */
+  static registry: FinalizationRegistry<RemoteQueryable> = new FinalizationRegistry((r_queryable: RemoteQueryable) => r_queryable.undeclare());
+
+  /** 
+   * @hidden
+   */
+  dispose() {
+    this.undeclare();
+    Queryable.registry.unregister(this);
+  }
+  /** 
+   * @hidden user must use new function
+   */
+  constructor(remote_queryable: RemoteQueryable, callback_queryable: boolean) {
+    this._remote_queryable = remote_queryable;
+    this._callback_queryable = callback_queryable;
+    Queryable.registry.register(this, remote_queryable, this)
+  }
+
+  /**
+   * receive next Query of this Queryable
+   * @returns Promise <Query | void>
+   */
+  async receive(): Promise<Query | void> {
+
+    if (this._callback_queryable === true) {
+      let message = "Cannot call `receive()` on Subscriber created with callback:";
+      console.log(message);
+      return
+    }
+
+    // QueryWS -> Query
+    let opt_query_ws = await this._remote_queryable.receive();
+    if (opt_query_ws != undefined) {
+      let query_ws = opt_query_ws[0];
+      let reply_tx = opt_query_ws[1];
+      return QueryWS_to_Query(query_ws, reply_tx);
+    } else {
+      console.log("Receieve returned unexpected void from RemoteQueryable");
+      return;
+    }
+  }
+
+  /**
+   * Undeclares Queryable
+   * @returns void
+   */
+  async undeclare() {
+    this._remote_queryable.undeclare();
+    Queryable.registry.unregister(this);
+  }
+
+  /** 
+   * @hidden : user must use declare_queryable
+   */
+  static new(remote_queryable: RemoteQueryable, callback_queryable: boolean): Queryable {
+    return new Queryable(remote_queryable, callback_queryable);
+  }
+}
+
+/**
+ * Convenience function to convert between QueryWS and Query 
+ * @hidden
+ */
+export function QueryWS_to_Query(
```

Comment:
```suggestion
function QueryWS_to_Query(
```

Reply:
`QueryWS_to_Query` is used inside `session.ts` so it needs to be exported.